### PR TITLE
Allow rpmbuild source fetching

### PR DESCRIPTION
**Pull request by k0ste**
Required for: https://github.com/netdata/netdata/pull/14551

### CODE REVIEW

File: package-builders/suse-build.sh [hunk 1]
```diff
@@ -29,6 +29,12 @@ cat > "/usr/src/packages/SOURCES/netdata-${pkg_version}/system/.install-type" <<
 	PREBUILT_DISTRO='${DISTRO} ${DISTRO_VERSION}'
 	EOF
 
+# Download Sources
+rpmbuild --nobuild --define "_topdir /root/rpmbuild/SOURCES" \
+--define "_sourcedir /root/rpmbuild/SOURCES" \
+--define "source_date_epoch_from_changelog false" \
+--undefine "_disable_source_fetch" "/root/rpmbuild/SPECS/netdata.spec" || exit 1
```
Comment (Ferroin) [inline package-builders/suse-build.sh:36]:
```suggestion
rpmbuild --nobuild --define "_topdir /usr/src/packages/SOURCES" \
--define "_sourcedir /usr/src/packages/SOURCES" \
--define "source_date_epoch_from_changelog false" \
--undefine "_disable_source_fetch" "/usr/src/packages/SPECS/netdata.spec" || exit 1
```

**Review by Ferroin — COMMENTED** [review]
This likely needs further modification so that it works correctly with the way we’re bypassing source downloading for the Netdata sources. See the comments starting at line 10 in each of these files, as well as the three `sed` invocations immediately following said comments. The reasoning may sound a bit silly, but it saves us 10-15 minutes _per build_ when building packages on platforms other than x86_64 (because we have to use QEMU userspace emulation for the builds).

Actually dealing with this will require either:
1. Updating that code to preserve the Source lines for the go plugin.
2. Updating that code to also fetch the Go sources outside of the RPM build process and put them in the correct locations like we’re doing for the main agent sources right now.

**Comment by k0ste** [comment]
As for me, the build is made very difficult. A bunch of Dockerfiles can be replaced with one [mock container](https://github.com/mmornati/docker-mock-rpmbuilder), passing to the container the desired distribution and architecture (x86/ARM builds already supported). The container will also pick up the sources on its own, in fact, this is a line from [there](https://github.com/mmornati/docker-mock-rpmbuilder/blob/master/build-rpm.sh#L135-L141)

**Comment by Ferroin** [comment]
> As for me, the build is made very difficult. A bunch of Dockerfiles can be replaced with one [mock container](https://github.com/mmornati/docker-mock-rpmbuilder), passing to the container the desired distribution and architecture (x86/ARM builds already supported). The container will also pick up the sources on its own, in fact, this is a line from [there](https://github.com/mmornati/docker-mock-rpmbuilder/blob/master/build-rpm.sh#L135-L141)

The problem is that this needs to run in CI _against the exact sources the CI job is running for_. Actually achieving that requires that we either:

- Create a source tarball just so that rpmbuild can turn around and extract it again. This is a waste of time in general, but as mentioned in my previous comment it adds a _huge_ overhead for non-x86 builds (10-15 minutes extra on average _per build_, which adds up to a couple of _hours_ longer for the workflow as a whole).
- Creatively convince rpmbuild that it’s already fetched and extracted the relevant sources by stripping out source entries and overriding the `%setup` macro so that it just copies the required data into the correct location.

To clarify, the issue here is not _fetching_ the tarballs, that’s easy to work around in numerous well documented ways. The issue is the wasted time from building a tarball just to have rpmbuild turn around and extract it again for a source tree that is already immediately available.

I understand that this leads to additional complexity, but it’s unfortunately not practical for us to change it significantly.

**Comment by k0ste** [comment]
Can we merge this to move the dependent PR in the main project?

**Comment by Ferroin** [comment]
As it is currently, it won’t actually work correctly to allow fetching the Go plugin sources. The points raised in https://github.com/netdata/helper-images/pull/204#pullrequestreview-1381751883 still apply.

**Comment by k0ste** [comment]
Can you be more specific about what is missing now because I can't guess

* the build is in different files, it's hard to put in one head and determine what's is going on
* no jobs for testing container builds via CI
* where to get the artifact before and after the modifications - with those specs, I'we could guess what it's all about

**Comment by Ferroin** [comment]
> Can you be more specific about what is missing now because I can't guess
> 
>     * the build is in different files, it's hard to put in one head and determine what's is going on

All the relevant code in this case is in `package-builders/fedora-build.sh` and `package-builders/suse-build.sh`. The only difference between the two scripts is the exact paths used, which are different between Fedora-like (CentOS, RedHat, Oracle, Amazon, Alma, etc) distros and SUSE distros.

The scripts expect the Netdata Agent sources (from https://github.com/netdata/netdata) to be mounted in the container at `/netdata`, and for a version to be defined using the `VERSION` environment variable.

The issue here is that we strip out _all_ `Source:` lines in the spec file in line 17 of these scripts, and override the `%setup` macro in line 18, with the specific goal of sidestepping rpmbuild’s insistence that it be given a compressed tarball of the agent sources instead of just using an existing directory (because jumping through the hoops required to use a tarball here actually has a major impact on how fast our builds run for architectures other than x86_64).

>     * no jobs for testing container builds via CI
>
>     * where to get the artifact before and after the modifications - with those specs, I'we could guess what it's all about

Indeed, this is a bit of a pain at times.

You can test locally though. See https://github.com/netdata/netdata/blob/master/packaging/building-native-packages-locally.md for info on how to build the packages locally in the same way we build them in CI. You can then substitute a local build of the relevant package builder container to test the changes made in this repo.

**Comment by k0ste** [comment]
Thanks @Ferroin 🙏
Now is clear, at least in a few hours I was able to make corrections. For now:

* update SPEC to be compatible with docker images
* verify local build for `almalinux8`
* verify local build for `tumbleweed`

```
Wrote: /root/rpmbuild/RPMS/aarch64/netdata-1.38.1-1.el8.aarch64.rpm
Wrote: /usr/src/packages/RPMS/aarch64/netdata-1.38.1-1.aarch64.rpm
```

**Comment by k0ste** [comment]
@Ferroin can you review please?

**Review by Ferroin — APPROVED** [review]

**Comment by Ferroin** [comment]
Holding off on merging this for the moment as we have a patch release scheduled for later this week, and we can’t currently isolate changes made in this repo from our release process.

**Comment by k0ste** [comment]
> Holding off on merging this for the moment as we have a patch release scheduled for later this week, and we can’t currently isolate changes made in this repo from our release process.

Seems `netdata` 1.39 was released, and plugin-split RPM changes was merged

**Pull request merged**